packages/gatsby-plugin-mdx/gatsby/create-schema-customization.js

```diff
@@ -93,20 +93,30 @@ module.exports = function createSchemaCustomization(
     }
   }
 
-  const processMDX = ({ node }) =>
-    genMDX({
-      node,
-      options,
-      store,
-      pathPrefix,
-      getNode,
-      getNodes,
-      cache,
-      reporter,
-      actions,
-      schema,
-      ...helpers,
-    })
+  const pendingPromises = new WeakMap()
+  const processMDX = ({ node }) => {
+    let promise = pendingPromises.get(node)
+    if (!promise) {
+      promise = genMDX({
+        node,
+        options,
+        store,
+        pathPrefix,
+        getNode,
+        getNodes,
+        cache,
+        reporter,
+        actions,
+        schema,
+        ...helpers,
+      })
+      pendingPromises.set(node, promise)
+      promise.then(() => {
+        pendingPromises.delete(node)
+      })
+    }
+    return promise
+  }
```

Comment:
This looks 👍 

Just wanted to ask if it wouldn't make sense to do so inside `getMDX` function - it already check for completed work cache in https://github.com/gatsbyjs/gatsby/blob/e1a1396e1582a18047e2cefaeff0f57d61c0c83a/packages/gatsby-plugin-mdx/utils/gen-mdx.js#L60-L68 (unless cache is disabled which we do during sourcing). This would then be familiar - "If already in progress -> return stored promise, if already done -> return cached result, else do actual work"

Reply:
I thought about this but the problem is that `genMdx` has lots of options and I am not confident enough about usage patterns for it. We call it with default options from `processMdx` - but what if somewhere else it is called with another set of options and the assumption about `payloadCacheKey` that works with the current set of use cases won't work anymore when this new codepath is added.

I guess you have more context, so if you think it will be OK - I can definitely change it.